### PR TITLE
[LLVMGPU] Create GPU pipeline option for IGEMM

### PR DESCRIPTION
**Pull request by Max191**
This PR moves the IGEMM transformation out of executable configuration, and creates a new GPU pipeline option for using IGEMM for convolutions. The PR reworks the way lowering configs are selected on convolutions, but the resulting configurations after the IGEMM transformation are unchanged. The PR makes the following changes:

- Creates a new field in GPUPipelineOptionsAttr called use_igemm_convolution. This new option indicates to the TileAndFuse pipeline to add the ConvolutionToIGEMM pass.
- Adds lowering config selection logic for convolutions going down the IGEMM path. The old way to select lowering configs was to transform into im2col + matmul, and then use the TileAndFuse matmul lowering config logic to set a configuration on the matmul. Now, the config is set on the convolution directly, and when the IGEMM transformation happens, the config is moved to the matmul.
- Refactor some of the TileAndFuse matmul config logic to reuse in the convolution config logic.
- Remove LLVMGPUConvolutionToIGEMMPass. This pass is no longer necessary, since we do not need a config function or control function.
- Removes various IGEMM related tests from the TileAndFuse test files, and creates new tests for the IGEMM path.
- Adds some util functions in LinalgExt to help infer the iteration domain of and mapping of the resulting matmul after the Conv2DToIm2col transformation. This is used when setting the lowering config on convolutions for IGEMM.

This change is largely NFC-ish. The main functional change here is that IGEMM will no longer be used when MFMA intrinsics cannot be targeted, but the IGEMM path is not performant when that is the case anyway. Otherwise, the changes from this PR simply change the ordering of how we select lowering strategies for IGEMM convolutions (and the pipeline options that are used for IGEMM). This will be useful for applying tuning configurations for IGEMM later on, since the transformation no longer runs during configuration.

### CODE REVIEW

**Comment by Max191** [comment]
I will double check everything is working on SDXL/iree-kernel-benchmark, but pipeline tests look good.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: We don't need a whole new pipeline for this, we can just call `ConvolutionToIGEMM` right at the start of `TileAndFuse` and we can control the application of im2col either with a pipeline option (probably not best but basically what this is) or with a lowering config attribute.

If you want to start with a pipeline option that works for me though.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]:
```suggestion
getMatmulLoweringConfigAndWorkgroupSize(SmallVector<int64_t> bounds,
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h:26]: We might not need this TODO actually. IGEMM might not be worth it without a special mma instruction.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Utils/Utils.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Utils/Utils.cpp]: nit: since we're type switching on named conv ops anyway, move these type getters inside the TypeSwitch and drop the `isaConvolutionOpInterface`

File: compiler/src/iree/compiler/Dialect/LinalgExt/Utils/Utils.h
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Utils/Utils.h]:
```suggestion
/// indexing. The resulting indexing maps should represent the indexing of some
```

**Review by qedawkins — CHANGES_REQUESTED** [review]
Nice, this LGTM with one main comment about not creating a new pipeline. After that looks good, thanks for the push!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: I started with the pipeline option, but then ran into all the python binding stuff that was recently added, and decided to just add a new pipeline. The nice thing about making it a new pass/pipeline option is that we can get rid of the control function, and the wrapper pass for LLVMGPU, so I don't really want to make it a lowering config attribute.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h:26]: I'll leave it for now and then we can reevaluate later whether we should do this.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: I've been trying to reduce the number of pipelines as pipelines have high maintenance cost and are much harder to configure as we try to fuse more both horizontally and vertically.

I see that this is basically a wrapper pipeline but we should be controlling applications of individual passes with pipeline options rather than new entry points. We've been planning to deduplicate LLVMGPUPadAndVectorDistribute for this reason for a while, and the rest of the compiler (Flow/Stream/HAL) all do this already.

FWIW since igemm is an algorithmic change that data seems to be showing we might not always want to do, I'm expecting we'll land on lowering config eventually anyway but I'm not blocking on that.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:90]: nit: use similar naming to the other flags: `iree-codegen-llvmgpu-test-igemm`. We're trying to keep dev flags prefixed with `experimental` or `test` until ready to turn on by default. It helps prevent proliferation of WIP flags. We can rename it to this once on by default.

**Review by qedawkins — APPROVED** [review]
🥳

**Comment by Max191** [comment]
Will wait for a review from @kuhar or @makslevental for the python bindings stuff.

**Review by MaheshRavishankar — APPROVED** [review]
Cool!

**Review by makslevental — APPROVED** [review]
Python pieces LGTM

**Pull request merged**
